cdap-docs/admin-manual/source/operations/logging.rst

```diff
@@ -1,44 +1,96 @@
-  directories under ``/var/run/cloudera-scm-agent/process``.
+- to be extensible using :ref:`custom log pipelines <logging-monitoring-custom-logging-pipeline>`; and
 
+- to use a system that can be adapted for individual requirements at both the application
+  and cluster levels.
 
-Logging Messages
-================
 
+Logging Example
+===============
+.. highlight:: console
+
+This diagram shows the steps CDAP follows when logging a program of an application:
+
+  - Logs are collected from an individual program running in a YARN container.
+
+  - YARN containers have their own local log files.
+  
+  - CDAP logs are published to Kafka.
+
+  - The Log Saver Service (*Log.saver*) is configured to read log messages from Kafka.
+    Log.saver reads the messages from Kafka, buffers and sorts them in-memory, before
+    persisting them to HDFS. These messages are written to files corresponding to the
+    programs the log messages belong to.
+
+  - Metrics are created based on the results of the pipeline, and passed to the Metrics Service.
+   
+  - For security, the files written out to persistent storage in HDFS have permissions set
+    such that they are accessible only by the ``cdap`` user.
+
+  - Log.saver also emits metrics for each *info, debug,* or *error* message received. This
+    helps in monitoring the amount of logs emitted by applications, application's programs,
+    system services, etc. and can be retrieved by querying the :ref:`CDAP metrics system
+    <http-restful-api-metrics>`.
+
+.. figure:: /_images/logging/logging-framework-simple.png
+    :figwidth: 100%
+    :width: 800px
+    :align: center
+
+    **CDAP Logging Example:** From a YARN container, through Kafka and the Log Saver Service, to HDFS
+
+Logging uses the standard `SLF4J (Simple Logging Facade for Java)
+<http://www.slf4j.org/manual.html>`__ APIs and `Logback
+<https://logback.qos.ch/manual>`__. Logging is configured using instances of Logback's
+"logback" file, consisting of *logging pipelines* with *logging appenders*:
+
+  - A **logging pipeline** is a process that consumes log events from Kafka, invokes
+    *logback appenders* defined in its configuration, buffering, sorting, and then
+    persisting the log messages.
+
+  - A **logback appender** (or *appender*) is a Java class, responsible for writing the
+    log events to persistent storage and maintaining metadata about that storage.
+
+
+User Application Program Logs
+=============================
+
+Emitting Log Messages from a Program
+------------------------------------
 .. highlight:: java
 
-CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
-<http://www.slf4j.org/manual.html>`__ APIs. For instance, in a flowlet you can write::
+CDAP supports application logging through the standard `SLF4J (Simple Logging Facade for Java)
+<http://www.slf4j.org/manual.html>`__ APIs.
+
+For instance, in a flowlet you can write::
```

Comment:
I think we should change this to debug instead of info (or perhaps even trace). We do not want to encourage excessive info logging.